### PR TITLE
Fix issue #125: sign block with old key if signing key updated in same block

### PR DESCRIPTION
**Pull request by abitmore**
PR for #125.

This PR also contains a fix about block size check, and a fix about removing item from fork db.

Things to be done:
- [x] test case for loading object_database from disk which contains signing keys not in genesis
- [x] test case for `witness_update_operation`
- [x] test case for `witness_create_operation`
- [x] <del>test case for `witness` plugin</del> update: I have done some manual tests, it's too much efforts to create such test cases.

### CODE REVIEW

**Comment by abitmore** [comment]
Travis complains:
>2795751ms th_a       object_database.cpp:106       open                 ] Opening object database from /tmp/graphene-tmp/e16c-550d-aba8-ade6 ...
chain_test: /home/travis/build/bitshares/bitshares-core/libraries/db/include/graphene/db/simple_index.hpp:67: const graphene::db::object& graphene::db::simple_index<T>::insert(graphene::db::object&&) [with T = graphene::chain::global_property_object]: Assertion `!_objects[instance]' failed.
Running 214 test cases...
unknown location(0): fatal error in "change_signing_key": signal: SIGABRT (application abort requested)
/home/travis/build/bitshares/bitshares-core/tests/common/database_fixture.cpp(271): last checkpoint

Related code:
https://github.com/bitshares/bitshares-core/blob/37e72adc4fa3f58efbb678972ab3c314107daa79/tests/tests/block_tests.cpp#L327-L331

Failed to reopen database file after closed? Any idea how to fix?

**Comment by pmconrad** [comment]
The database is not meant to be reopened after close. Also see discussion in #436 #687 #689 .

**Comment by abitmore** [comment]
Destroyed db object and created a new one. Travis-CI no longer complains.

File: libraries/plugins/witness/witness.cpp [hunk 1]
```diff
@@ -250,7 +253,7 @@ block_production_condition::block_production_condition_enum witness_plugin::mayb
    }
 
    fc::time_point_sec scheduled_time = db.get_slot_time( slot );
-   graphene::chain::public_key_type scheduled_key = scheduled_witness( db ).signing_key;
+   graphene::chain::public_key_type scheduled_key = *db.find_witness_key_from_cache( scheduled_witness ); // should be valid
```
Comment (pmconrad) [inline libraries/plugins/witness/witness.cpp]: I think instead of using the cache, witness_plugin could simply connect to the ``database::applied_block`` signal and use that to retrieve the latest keys of the witness(es) it is interested in.

File: libraries/chain/db_block.cpp [hunk 2]
```diff
@@ -331,10 +331,19 @@ signed_block database::_generate_block(
    witness_id_type scheduled_witness = get_scheduled_witness( slot_num );
    FC_ASSERT( scheduled_witness == witness_id );
 
-   const auto& witness_obj = witness_id(*this);
-
    if( !(skip & skip_witness_signature) )
-      FC_ASSERT( witness_obj.signing_key == block_signing_private_key.get_public_key() );
+   {
+      auto signing_key = find_witness_key_from_cache( witness_id );
```
Comment (pmconrad) [inline libraries/chain/db_block.cpp]: Instead of using the cache, simply move the existing FC_ASSERT down a couple of lines, after ``_pending_tx_session.reset();``

File: libraries/plugins/witness/witness.cpp (hunk 1 above)
Comment (abitmore) [reply libraries/plugins/witness/witness.cpp]: Essentially you mean to move the cache to witness plugin. Because there is no signal when a block is popped out, in order to deal with chain reorganization, the cache should be in (extended) chain state. Then, in order to make it work, need to either replay the whole chain to enable the plugin, or add code to refresh the cache (copy data from main chain state to the extended part) on startup. Then, in case when there is a reorganization to the startup head block, need to make sure the cache don't get emptied.

I didn't see it's simpler so far.

File: libraries/chain/db_block.cpp (hunk 2 above)
Comment (abitmore) [reply libraries/chain/db_block.cpp]: It's cheaper to check here than checking after called `_pending_tx_session.reset();`.

Comment (abitmore) [reply libraries/chain/db_block.cpp]: Even for nodes with proper keys, it's still slightly cheaper because the cache would be much smaller than object database.

File: libraries/plugins/witness/witness.cpp (hunk 1 above)
Comment (abitmore) [reply libraries/plugins/witness/witness.cpp]: I agree that ideally it's better to put the logic in witness plugin because block generation is actually witness plugin's business.

> Blocks are only popped when switching forks, which means there will be block pushes immediately after the pop, which means the cache will automatically be updated in that case.

One scenario to consider: if put the cache in witness plugin, if we code the plugin to track applied operations only, after a block is popped out, the cache may become inconsistent if the popped out block contains a `witness_update_operation`, when pushing a new block, if there is no the same `witness_update_operation` in the new block, the cache would still be inconsistent. To solve this, we need additional code to detect if a chain reorganization has occurred and refresh the cache accordingly, if not simply refresh the whole cache on every block. However, when refreshing the cache, we need to fetch data from main chain state, since signalling is a multi-threading thing, the main state could have changed when fetching data, which means the cache is not 100% reliable.

By putting the cache in database class we can guarantee its consistency. Perhaps there will still be edge cases, for example, witness plugin tries to generate a block in the middle when database is pushing a block.

Another thing is we don't have unit test for witness plugin, so adding more code there means more work / risks.

Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: The witness plugin *typically* needs to keep track of the keys of only one witness. Refreshing this after each pushed block should have minimal cost, and only for the node in question.

Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: I think the applied_block signal is processed in sync right after the block has been applied. There shouldn't be any tx's pushed in between.
https://github.com/bitshares/bitshares-core/blob/2.0.180612/libraries/chain/db_block.cpp#L550

Comment (abitmore) [reply libraries/plugins/witness/witness.cpp]: Oh, right, I forgot that. The state should be clean when processing applied_block event.

By the way, why we need to sleep here and there in test cases? IIRC sometimes chain_test will segfault without sleeping. E.G. https://github.com/bitshares/bitshares-core/blob/95906d8100198a070ab1894e465a8611bc559e56/tests/tests/market_rounding_tests.cpp#L76

Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: I think it's required in some places where API callbacks happen. These are asynchronous.
Shouldn't be necessary in market_rounding_tests IMO.

File: libraries/chain/db_block.cpp (hunk 2 above)
Comment (pmconrad) [reply libraries/chain/db_block.cpp]: For nodes with keys it may be cheaper, but the global cache is a penalty for *all* nodes.

**Comment by abitmore** [comment]
Moved the signing key cache to witness plugin as recommended by @pmconrad.

File: libraries/chain/db_block.cpp [hunk 3]
```diff
@@ -377,17 +366,40 @@ signed_block database::_generate_block(
...
+
+   // pop pending state (reset to head block state)
    _pending_tx_session.reset();
+
+   // Check witness signing key
+   if( !(skip & skip_witness_signature) )
+   {
+      // Note: if this check failed (which won't happen in normal situations),
+      // we would have temporarily broken the invariant that
+      // _pending_tx_session is the result of applying _pending_tx.
```
Comment (pmconrad) [inline libraries/chain/db_block.cpp]: At this point _pending_tx_session has been cleared, so the comment makes no sense.

File: libraries/chain/db_block.cpp [hunk 4]
```diff
@@ -532,7 +538,17 @@ void database::_apply_block( const signed_block& next_block )
    uint32_t skip = get_node_properties().skip_flags;
    _applied_ops.clear();
 
-   FC_ASSERT( (skip & skip_merkle_check) || next_block.transaction_merkle_root == next_block.calculate_merkle_root(), "", ("next_block.transaction_merkle_root",next_block.transaction_merkle_root)("calc",next_block.calculate_merkle_root())("next_block",next_block)("id",next_block.id()) );
+   if( !(skip & skip_block_size_check) )
```
Comment (pmconrad) [inline libraries/chain/db_block.cpp]: Should at least soft-fork protect this.

File: tests/tests/block_tests.cpp [hunk 5]
```diff
@@ -1726,4 +1843,90 @@ BOOST_FIXTURE_TEST_CASE( temp_account_balance, database_fixture )
...
+         maybe_large_block.timestamp = db.get_slot_time(1);
+         maybe_large_block.transaction_merkle_root = maybe_large_block.calculate_merkle_root();
+         maybe_large_block.witness = db.get_scheduled_witness(1);
+         maybe_large_block.sign(key);
+         idump( (fc::raw::pack_size(maybe_large_block)) );
+
+         // should fail to push if it's too large
+         if( fc::raw::pack_size(maybe_large_block) > gpo.parameters.maximum_block_size )
+         {
+            BOOST_CHECK_THROW( db.push_block(maybe_large_block), fc::exception );
```
Comment (pmconrad) [inline tests/tests/block_tests.cpp]: The test should ensure that this is triggered a couple of times. Perhaps increment a counter each time and BOOST_CHECK_EQUAL the counter after the loop ends. (Not sure how often it should trigger...)

File: libraries/chain/db_block.cpp (hunk 3 above)
Comment (abitmore) [reply libraries/chain/db_block.cpp]: Yes, `_pending_tx_session` has been cleared, but `_pending_tx` hasn't, so they're inconsistent.

File: tests/tests/block_tests.cpp (hunk 5 above)
Comment (abitmore) [reply tests/tests/block_tests.cpp]: Added a check to make sure it has been triggered at least once. I think it's enough. Didn't use `BOOST_CHECK_EQUAL` because it would be a pain when trying to tweak the loop in the future.
